Improbable/Schema/Improbable.Schema.Bundle/FieldDefinitionExtensions.cs

```diff
@@ -0,0 +1,122 @@
+                FieldType.Map => throw new InvalidOperationException("Invalid for the map type. Check the key and value types individually."),
+                FieldType.Singular => field.SingularType.Type.HasEnum(),
+                _ => throw new ArgumentOutOfRangeException()
+            };
+        }
+
+        public static bool HasPrimitive(this FieldDefinition field)
+        {
+            return field.TypeSelector switch
+            {
+                FieldType.Option => field.OptionType.InnerType.HasPrimitive(),
+                FieldType.List => field.ListType.InnerType.HasPrimitive(),
+                FieldType.Map => throw new InvalidOperationException("Invalid for the map type. Check the key and value types individually."),
+                FieldType.Singular => field.SingularType.Type.HasPrimitive(),
+                _ => throw new ArgumentOutOfRangeException()
+            };
+        }
+
+        public static bool HasPrimitive(this FieldDefinition field, PrimitiveType type)
+        {
+            return field.TypeSelector switch
+            {
+                FieldType.Option => field.OptionType.InnerType.HasPrimitive(type),
+                FieldType.List => field.ListType.InnerType.HasPrimitive(type),
+                FieldType.Map => throw new InvalidOperationException("Invalid for the map type. Check the key and value types individually."),
+                FieldType.Singular => field.SingularType.Type.HasPrimitive(type),
+                _ => throw new ArgumentOutOfRangeException()
+            };
+        }
+
+        public static bool HasCustomType(this FieldDefinition field)
+        {
+            return field.TypeSelector switch
+            {
+                FieldType.Option => field.OptionType.InnerType.HasCustomType(),
+                FieldType.List => field.ListType.InnerType.HasCustomType(),
+                FieldType.Map => throw new InvalidOperationException("Invalid for the map type. Check the key and value types individually."),
+                FieldType.Singular => field.SingularType.Type.HasCustomType(),
+                _ => throw new ArgumentOutOfRangeException()
+            };
+        }
+
+        public static bool HasCustomType(this FieldDefinition field, string qualifiedTypeName)
+        {
+            return field.TypeSelector switch
+            {
+                FieldType.Option => field.OptionType.InnerType.HasCustomType(qualifiedTypeName),
+                FieldType.List => field.ListType.InnerType.HasCustomType(qualifiedTypeName),
+                FieldType.Map => throw new InvalidOperationException("Invalid for the map type. Check the key and value types individually."),
+                FieldType.Singular => field.SingularType.Type.HasCustomType(qualifiedTypeName),
+                _ => throw new ArgumentOutOfRangeException()
+            };
+        }
+
+        public static bool CanPrimitiveBeNull(this FieldDefinition field)
```

Comment:
Can `Entity` be null? Or are you not supporting that, I can't recall

Reply:
Ah nope, not supported